src/Map.ts

```diff
@@ -0,0 +1,465 @@
+ * true, wrapped in a `Some`. If no key is found, returns `None`. Uses
+ * the given `OrderingComparer` if passed, otherwise defaults to default
+ * ASCII-based sort.
+ *
+ * @category Lookups
+ */
+const findKey =
+    <K extends {}>(
+        predicate: Predicate<K>,
+        orderingComparer: OrderingComparer<K> = defaultOrderingComparer
+    ) =>
+    <V>(map: ReadonlyMap<K, V>): Option<K> =>
+        Option.ofNullish(keys(orderingComparer)(map).find(predicate))
+
+/**
+ * Creates a new empty map. Essentially an alias for `new globalThis.Map()`.
+ * Provided for convience to avoid having to use `globalThis`.
+ *
+ * @category Constructors
+ */
+const empty = <K = never, V = never>() => new globalThis.Map<K, V>()
+
+/**
+ * Returns `true` if at least one _value_ in the `Map` returns `true`
+ * for the given predicate function.
+ *
+ * @category Lookups
+ * @category Utils
+ */
+const exists =
+    <V>(predicate: Predicate<V>) =>
+    <K>(map: ReadonlyMap<K, V>): boolean => {
+        if (map.size < 1) {
+            return false
+        }
+
+        for (const [, v] of map) {
+            if (predicate(v)) {
+                return true
+            }
+        }
+
+        return false
+    }
+
+/**
+ * Replace the value at a given key in the map using the given
+ * replacement function. Will use the given `EqualityComparer`
+ * if passed. Otherwise defaults to reference equality (triple equals).
+ *
+ * If the key isn't in the map, returns the map unchanged.
+ *
+ * @category Transformations
+ */
+const change =
+    <K, V>(
+        key: K,
+        f: (v: V) => V,
+        equalityComparer: EqualityComparer<K> = defaultEqualityComparer
+    ) =>
+    (map: ReadonlyMap<K, V>): ReadonlyMap<K, V> =>
+        pipe(
+            map,
+            findWithKey(key, equalityComparer),
+            Option.match({
+                some: ([k, v]) => {
+                    const copy = new globalThis.Map(map)
+                    copy.set(k, f(v))
+                    return copy
+                },
+                none: map,
+            })
+        )
+
+/**
+ * Get the number of key/value pairs in the map.
+ *
+ * @category Utils
+ */
+const size = <K, V>(map: ReadonlyMap<K, V>) => map.size
+
+/**
+ * Returns whether the map contains any key/value pairs.
+ *
+ * @category Utils
+ *
+ * @returns `true` if the map has no bindings, `false` otherwise.
+ */
+const isEmpty = <K, V>(map: ReadonlyMap<K, V>) => map.size < 1
+
+/**
+ * Get only the keys from the map as an array. Will use the given `OrderingComparer`
+ * to sort the keys, otherwise will the default ASCII-based sort.
+ *
+ * @category Utils
+ */
+const keys =
+    <K>({ compare }: OrderingComparer<K> = defaultOrderingComparer) =>
+    <V>(map: ReadonlyMap<K, V>): readonly K[] =>
+        Array.from(map.keys()).sort(compare)
+
+/**
+ * Gets all the values from the map as an array, including duplicates. Values
+ * will be sorted by their corresponding key, using the default ASCII-based
```

Comment:
> Values will be sorted by their corresponding key

Are we sure this is expected?  I would have expected passing in a `OrderingComparer` to a function called `values` would sort by the values, not the keys. (But feel free to ignore me there's precedence I'm missing)

Reply:
I actually went back and forth on this as well... the F# `Map` api sorts by key, the fp-ts `Map` api sorts by value 🤷‍♂️ Feels a little bit 6-one-way-half-dozen-the-other. I guess it feels slightly more consistent to the api to sort by key since this would be the only place in this api where we would use an `OrderingComparer<V>` instead of `OrderingComparer<K>` 🤔

Reply:
>Feels a little bit 6-one-way-half-dozen-the-other.

We're fairly significantly changing the result though, right? It feels like if we aren't even getting the keys back, how does a consumer even know it is sorted? 😅

Reply:
So I checked and fp-ts actually does take an `Ord<V>` and not an `Ord<K>` so I think you're right--it makes sense to do the sort by value, not by key

Reply:
https://github.com/taylorn16/fp-toolkit/pull/20/commits/618c38033a54125fcb8de05034f145c33c628f8b